File: Day1.hs
module Day1 where

main :: IO ()
main = do
  masses <- readMasses
  let totalMass = sum masses
  let totalFuel = sum $ map fuel $ masses
  let totalFuel' = sum $ map fuel' $ masses
  print $ totalFuel
  print $ totalFuel' - totalMass

readMasses :: IO [Int]
readMasses = map read . lines <$> readFile "Day1.txt"

fuel :: Int -> Int
fuel mass = mass `div` 3 - 2

fuel' :: Int -> Int
fuel' mass = go mass 0
  where
    go :: Int -> Int -> Int
    go mass acc
      | mass > 0 = go (fuel mass) (acc + mass)
      | otherwise = acc

File: Day1.py
def fuel(mass):
    return mass // 3 - 2


def fuel_(mass):
    total = 0
    while mass > 0:
        total += mass
        mass = fuel(mass)
    return total

def read_masses():
    contents = open("Day1.txt", "r").read().strip()
    lines = contents.split("\n")
    weights = [int(line) for line in lines]
    return weights


if __name__ == "__main__":

    masses = read_masses()

    print(sum(map(fuel, masses)))

    # To get the fuel mass only, we subtract the mass afterwards
    print(sum(map(fuel_, masses)) - sum(masses))
